sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Shard.java

```diff
@@ -0,0 +1,97 @@
+ *
+ * The log files in each shardPath are guaranteed to contain mutually exclusive blobs, and can be consumed and
+ * processed in parallel without violating the ordering of modifications per blob during the iteration.
+ */
+class Shard  {
+
+    private final BlobContainerAsyncClient client; /* Changefeed container */
+    private final String shardPath; /* Shard virtual directory path/prefix. */
+    private final ChangefeedCursor segmentCursor; /* Cursor associated with parent segment. */
+    private final ChangefeedCursor userCursor; /* User provided cursor. */
+    private final ChunkFactory chunkFactory;
+
+    /**
+     * Creates a new Shard.
+     */
+    Shard(BlobContainerAsyncClient client, String shardPath, ChangefeedCursor segmentCursor,
+        ChangefeedCursor userCursor, ChunkFactory chunkFactory) {
+        this.client = client;
+        this.shardPath = shardPath;
+        this.segmentCursor = segmentCursor;
+        this.userCursor = userCursor;
+        this.chunkFactory = chunkFactory;
+    }
+
+    /**
+     * Get events for the Shard.
+     * @return A reactive stream of {@link BlobChangefeedEventWrapper}
+     */
+    Flux<BlobChangefeedEventWrapper> getEvents() {
+        /* List relevant chunks. */
+        return listChunks()
+            .concatMap(chunkPath -> {
+                /* Defaults for blockOffset and objectBlockIndex. */
+                long blockOffset = 0;
+                long objectBlockIndex = 0;
+                /* If a user cursor was provided and it points to this chunk path, the chunk should get events based
+                   off the blockOffset and objectBlockIndex.
+                   This just makes sure only the targeted chunkPath uses the blockOffset and objectBlockIndex to
+                   read events. Any subsequent chunk will read all of its events (i.e. blockOffset = 0). */
+                if (userCursor != null && userCursor.getChunkPath().equals(chunkPath)) {
+                    blockOffset = userCursor.getBlockOffset();
+                    objectBlockIndex = userCursor.getObjectBlockIndex();
+                }
+                return chunkFactory.getChunk(chunkPath, segmentCursor.toChunkCursor(chunkPath),
+                    blockOffset, objectBlockIndex)
+                    .getEvents();
+            });
+    }
+
+    /**
+     * Lists relevant chunks in a shard.
+     * @return A reactive stream of chunks.
+     */
+    private Flux<String> listChunks() {
+        Flux<String> chunks = client.listBlobs(new ListBlobsOptions().setPrefix(shardPath))
+            .map(BlobItem::getName);
+        /* If no user cursor was provided, just return all chunks without filtering. */
+        if (userCursor == null) {
+            return chunks;
+        /* If a user cursor was provided, filter out chunks that come before the chunk specified in the cursor. */
+        } else {
+            AtomicBoolean pass = new AtomicBoolean(); /* Whether or not to pass the event through. */
```

Comment:
this `pass` var get shared across all the subscribers of the returned `Flux<String>` instance, if that is not intended then we can wrap filter in a defer.

Reply:
fixed.